lib/dao.js

```diff
@@ -624,7 +624,193 @@ DataAccessObject.upsert = function(data, options, cb) {
+              data: update,
+              currentInstance: inst,
+              hookState: ctx.hookState,
+              options: options,
+            };
+            Model.notifyObserversOf('persist', context, function(err) {
+              if (err) return done(err);
+              connector.upsertWithWhere(modelName, where, update, options, done);
+            });
+          }
+          function done(err, data, info) {
+            if (err) return cb(err);
+            var context = {
+              Model: Model,
+              data: data,
+              isNewInstance: info && info.isNewInstance,
+              hookState: ctx.hookState,
+              options: options,
+            };
+            Model.notifyObserversOf('loaded', context, function(err) {
+              if (err) return cb(err);
+              var obj;
+              if (data && !(data instanceof Model)) {
+                inst._initProperties(data, { persisted: true });
+                obj = inst;
+              } else {
+                obj = data;
+              }
 
+              var context = {
+                Model: Model,
+                instance: obj,
+                isNewInstance: info ? info.isNewInstance : undefined,
+                hookState: hookState,
+                options: options,
+              };
+              Model.notifyObserversOf('after save', context, function(err) {
+                cb(err, obj);
+              });
+            });
+          }
+        });
+      } else {
+        var element = Object.keys(where)[0];
+        var filter = { where: where };
+        if (!Object.keys(data).length) {
+          var err = new Error('data object cannot be empty!');
+          err.statusCode = 400;
+          process.nextTick(function() { cb(err); });
+        }
+
+        if (!data.hasOwnProperty(element)) {
+          data[element] = where[element];
+        }
+        var primaryKey = idName(Model);
+        self.find(filter, function(err, instances) {
+          if (err) cb(err);
+          else {
+            var modelsLength = instances.length;
+            if (modelsLength === 0 && data.hasOwnProperty(primaryKey))  {
+              return self.create(data, options, cb);
+            } else if (modelsLength == 0 && data.hasOwnProperty(primaryKey) == false) {
+              var error = new Error('Error message: No matching instance found.Found error while trying ' +
+                'to create a new instance.' + 'Primary key and value ' + 'is not provided in the payload');
+              callback(error);
+            } else if (modelsLength === 1) {
+              var modelInst = instances[0];
+              return modelInst.updateAttributes(data, options, cb);
+            } else {
+              process.nextTick(function() {
+                var error = new Error('There are multiple instances found.' +
+                  'Upsert Operation will not be performed');
+                cb(error);
+              });
```

Comment:
@Amir-61 : 
I have put a note for this in my comment yesterday. Guess you missed it.  Putting it here again :  
Note:
I had to revert one change where I was checking if the data had primary key element or not, which I had done in the previous commit while addressing one of your comments. But the flow failed while entering find operation itself, with the foll error : so, giving "id":"" in the data in case of auto-generated ids helps.

/Users/space/Documents/office/demoSF/node_modules/loopback-connector/lib/connector.js:143
return this.getDataSource(model).idName(model);
^

TypeError: Cannot read property 'idName' of undefined
at Memory.Connector.idName (/Users/space/Documents/office/demoSF/node_modules/loopback-connector/lib/connector.js:143:35)
at Memory.Connector.getIdValue (/Users/space/Documents/office/demoSF/node_modules/loopback-connector/lib/connector.js:206:28)

Due to the time constraints we have at our end, I would like to keep the primaryKey check in the data payload, and take it up as incremental fix after the merge happens. The DB2 connector we have also behaves the same way. So, its acceptable as connector users from our end. We have a hard stop on tuesday (july19) for getting this merged since a lot of epics are just waiting for this feature to come in and it will have a ripple effect on all of them if not done on time. The current state of code meets all that our requirement needs and I believe to have addressed most of your comments. However, I can give incremental fixes by creating issues which you think should be addressed as future extensions once the code is merged like its being done for all other modules.